### PR TITLE
Check against payloads with version ids that do not match context version id

### PR DESCRIPTION
**Pull request by JoshuaBehrens**
### 1. Why is this change necessary?

There is a plugin in the community store that clones an entity and updates the cloned entity with a payload that still refers to the live id but has a clone version id in the context. This is actually the plugins fault and I contacted the manufacturer about it. We cool. But how to prevent this early on?

Now the existence prefetch ONLY LOADS EXISTENCE WITH CONTEXT VERSION ID [see here](https://github.com/shopware/platform/pull/2646/files#diff-7f2b0ab91c268388107b9e5b64f05e9316ce4b5158ece1d6463d195ce1fe40f9L372-L375). But the existence checks VALIDATE EXISTENCE WITH PAYLOAD VERSION ID. So we have a conflict there resulting in a mismatch of "Do I need to update or create?".

I see two ways to fix:
1. Trust the payload. Prefetch with the payload version ids is easy to do, not that bad actually, but as Shopware expects at a lot of places to expect the version id in the context is right, we should not go this way.
2. This PR to trust the context. And if the context goes: well we have this version id, then I 

### 2. What does this change do, exactly?

Checks whether version ids are given in payload and flip tables, when they do not match the context version id.

### 3. Describe each step to reproduce the issue or behaviour.

1. Clone an entity to get a new version id
2. Use payload with live version id to run an update in a version scoped context
3. Get an exception that you run update but Shopware tries to insert
4. Look into DB with values from the payload
5. They exist. Why does Shopware THINK IT NEEDS TO CREATE?!

### 4. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/workflow/2020-08-03-implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### 5. I need feedback

What does Shopware think about that? Am I approaching the issue correctly? I know, tests missing, changelog missing, exception looks like an untested IED. But I don't know what to do with it? If this plugin can fail, anyone working with the API can fail.

<a href="https://gitpod.io/#https://github.com/shopware/platform/pull/2646"></a>

### CODE REVIEW

File: src/Core/Framework/DataAbstractionLayer/Dbal/EntityWriteGateway.php [hunk 1]
```diff
@@ -348,9 +350,14 @@ private function prefetch(EntityDefinition $definition, array $pks, WriteParamet
                     if ($id === null) {
                         continue 2;
                     }
+
                     $newIds[] = Uuid::fromHexToBytes($id);
                 }
 
+                if ($versionField instanceof VersionField && ($pk[$versionField->getPropertyName()] ?? null) !== $contextVersionId) {
+                    throw new \LogicException('We have a version id in the context and a payload that states a different version id. WTF');
```
Comment (keulinho) [inline src/Core/Framework/DataAbstractionLayer/Dbal/EntityWriteGateway.php]: Please rename the error message to something more serious :rofl: 

and i think it would help to print both version numbers inside the error message to make debugging easier :+1:

**Review by keulinho — COMMENTED** [review]
Nice one! I guess it was pretty hard figuring out the root cause of the issue, checking that inside shopware definetly makes sense

File: src/Core/Framework/DataAbstractionLayer/Dbal/EntityWriteGateway.php (hunk 1 above)
Comment (JoshuaBehrens) [reply src/Core/Framework/DataAbstractionLayer/Dbal/EntityWriteGateway.php]: Hi @keulinho 

as mentioned in the ticket: I know I need to polish it :D

I just had a call with @shyim . We are at the point that we trigger this with a feature deprecation on 6_5_0_0. At normalization step, I will remove the version fields from the payload so the context wins. This does not make the version lookup more correct but it removes the issuing data.

YES it was a hell to find that. A combination of: why are we on a create? where is the existence check set in all these calls? And why the hell is the version_id in the payload different to the version id in the context? The last one is where you start to question A LOT.

**Comment by JoshuaBehrens** [comment]
I received feedback (see comments). I improved deprecation message, added a test for the removal, added a changelog. I did not add a test for the feature check because I don't know how to test it. It is also now very unlikely to happen due to the removal. Anyway this is wrong and needs to be validated and therefore it should stay there.

**Review by keulinho — APPROVED** [review]

File: src/Core/Framework/DataAbstractionLayer/Write/WriteCommandExtractor.php [hunk 2]
```diff
@@ -49,7 +50,10 @@ class WriteCommandExtractor
 {
     private EntityWriteGatewayInterface $entityExistenceGateway;
 
-    private DefinitionInstanceRegistry $definitionRegistry;
+    /**
+     * @var DefinitionInstanceRegistry
```
Comment (dneustadt) [inline src/Core/Framework/DataAbstractionLayer/Write/WriteCommandExtractor.php]: Why is that change necessary?

Comment (JoshuaBehrens) [reply src/Core/Framework/DataAbstractionLayer/Write/WriteCommandExtractor.php]: Oh sorry. Is is not. I just copied my changes from vendor to here and likely you changed this quite recently. Will amend

File: src/Core/Framework/DataAbstractionLayer/Dbal/EntityWriteGateway.php [hunk 3]
```diff
@@ -348,9 +350,19 @@ private function prefetch(EntityDefinition $definition, array $pks, WriteParamet
                     if ($id === null) {
                         continue 2;
                     }
+
                     $newIds[] = Uuid::fromHexToBytes($id);
                 }
 
+
```
Comment (dneustadt) [inline src/Core/Framework/DataAbstractionLayer/Dbal/EntityWriteGateway.php]: Sorry, this PR somehow slipped through the cracks 😬 So we can proceed, please remove this extra line break.

File: src/Core/Framework/Test/DataAbstractionLayer/Write/WriteCommandExtractorTest.php [hunk 4]
```diff
@@ -8,8 +8,12 @@
 use Shopware\Core\Framework\Context;
 use Shopware\Core\Framework\DataAbstractionLayer\Dbal\EntityDefinitionQueryHelper;
 use Shopware\Core\Framework\DataAbstractionLayer\DefinitionInstanceRegistry;
+use Shopware\Core\Framework\DataAbstractionLayer\Write\Command\WriteCommandQueue;
+use Shopware\Core\Framework\DataAbstractionLayer\Write\EntityWriteGatewayInterface;
```
Comment (dneustadt) [inline src/Core/Framework/Test/DataAbstractionLayer/Write/WriteCommandExtractorTest.php]: Please remove this unused import.

**Comment by JoshuaBehrens** [comment]
Rebased and solved the given remarks

**Comment by dneustadt** [comment]
Please have a look at the [tests that failed](https://github.com/shopware/platform/actions/runs/3285870844/jobs/5413375138) in the pipeline. These changes should be behind the major feature flag `v6.5.0.0`, since they seem to be breaking. If the tests are expected to fail after the feature flag would be removed, you may skip them accordingly: `Feature::skipTestIfActive('v6.5.0.0', $this);`

**Comment by shyim** [comment]
I will close this PR as it's inactive for a while.
Feel free to create a new one
